frontend/views/utils/avatar.js

```diff
@@ -0,0 +1,26 @@
+import sbp from '~/shared/sbp.js'
+import blockies from '@utils/blockies.js'
+import Colors from '@model/colors.js'
+import { randomFromArray } from '~/frontend/utils/giLodash.js'
+import { imageDataURItoBlob, imageUpload } from '@utils/image.js'
+
+const colorOptions = ['primary', 'warning', 'danger', 'success']
+const theme = Colors.light
+
+sbp('sbp/selectors/register', {
+  'gi.utils/avatar/create': function () {
+    // creates a random avatar image as a blob
+
+    const palette = randomFromArray(colorOptions)
+    const avatarBase64 = blockies.create({
+      bgcolor: theme[`${palette}_0`], // darkest
+      color: theme[`${palette}_1`], // medium
+      spotcolor: theme[`${palette}_2`], // lightest
+      size: 6,
+      scale: 12
+    }).toDataURL('image/png')
+    const avatarBlob = imageDataURItoBlob(avatarBase64)
+
+    return imageUpload(avatarBlob)
```

Comment:
Nice, this PR looks like it has simplified and cleaned up the code a good amount!

Question, this code (1) used to have a try/catch around it, and (2) is currently called in the following manner: `const picture = await sbp('gi.utils/avatar/create') || defaultAvatar` — so, are you ensuring that if  there is an error anywhere here things will be properly handled?

Reply:
Thanks for catching it. I made a change where in case avatar creation goes wrong the avatar sbp function returns null. That way, the default avatar is used if an error occurs. ( `const picture = await sbp('gi.utils/avatar/create') || defaultAvatar` )

I checked if it happens by manually throwing an error inside the `try {}` block. But pls let me know if you think there's a better way.

Reply:
@SebinSong Here's the thing... look through the sequence of events and possible errors in the line you quoted: `const picture = await sbp('gi.utils/avatar/create') || defaultAvatar`

Ask yourself the following questions:

Can something else go wrong?

Yes, in fact it can: `await sbp('gi.utils/avatar/create')` can throw an exception, because the promise that `imageUpload` returns can be rejected.

Do we handle that?

No.

How should we handle that?

Well, think about this from the perspective of the user and go through the workflow step-by-step:

1. This code gets called if we're signing up a new user.
2. If an exception occurs on `imageUpload`, it could be due to a variety of situations. Enumerate them all:
  1. The user supplied their own picture and it was too large, so the server rejected it. (*Oops!* Looks like we already caught a bug! This PR seems to have removed the ability for a user to supply their own `picture`?  Line 11 in `frontend/controller/actions/identity.js`?)
     - How should this situation be handled? Is it currently handled? Should the signup fail with a user-facing message? Should we proceed with a generated avatar? What do you think would be the behavior that made the most sense to you if you were signing up and this happened? How should the app handle it? Does it currently do that? If not, can you make it behave the ideal way?
  2. `imageUpload` failed for some other reason. Perhaps the server is down? How are we handling this right now? If you decide to simply use `defaultAvatar`, how can you do that if the way the current code happens it won't ever reach the `|| defaultAvatar` because an exception was thrown because a promise was rejected?

Since `imageUpload` is outside of the `try/catch`, we encounter problems like this. Should it be outside of it? If so, then the caller should also have an appropriate `try/catch` to handle these scenarios. I encourage you to really think through as a user using the app. What do you think should happen in all of these situations? Then make sure the code does it.

So these are some of the issues that jumped out at me.

Reply:
@taoeffect Thanks for the review.
Just put back the `picture` argument to the `gi.actions/identity/create` sbp function and make the random avatar generation happen in `gi.actions/identity/signup` as we discussed. This way, the both error cases you mentioned above are handled.

1. in case random avatar generation goes wrong: `gi.utils/avatar/create` returns `null` and the default avatar will be used.
2. in case imageUpload fails: `new GIErrorUIRuntimeError` will be thrown and the user will be presented with the error message saying what's gone wrong.

Pls let me know if there should be additional steps taken.